infra/RepoManager.py

```diff
@@ -0,0 +1,202 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Class to manage a git repository via python.
+
+This class is to be used to implement git commands over
+a python API and manage the current state of the git repo.
+
+  Typical usage example:
+
+    r_man =  RepoManager('https://github.com/google/oss-fuzz.git')
+    r_man.checkout('5668cc422c2c92d38a370545d3591039fb5bb8d4')
+"""
+import logging
+import os
+import shutil
+import subprocess
+
+
+class RepoManagerError(Exception):
+  """Class to describe the exceptions in RepoManager."""
+
+  def __init__(self, message):
+    """ Init the exception class.
+
+    Args:
+      message: the message to propagate to user
+    """
+    super().__init__(message)
+
+
+class RepoManager(object):
+  """Class to manage git repos from python.
+
+  Attributes:
+    repo_url: The location of the git repo
+    base_dir: The location of where the repo clone is stored locally
+    repo_name: The name of the github project
+    repo_dir: The location of the main repo
+  """
+
+  def __init__(self, repo_url, base_dir):
+    """Constructs a repo manager class.
+
+    Args:
+      repo_url: The github url needed to clone
+      base_dir: The full filepath where the git repo is located
+    """
+
+    self.repo_url = repo_url
+    self.base_dir = base_dir
+    self.repo_name = self.repo_url.split('/')[-1].strip('.git')
+    self.repo_dir = os.path.join(self.base_dir, self.repo_name)
+    self._clone()
+
+  def _clone(self):
+    """Creates a clone of the repo in the specified directory.
+
+      Raises:
+        RepoManagerError if the repo was not able to be cloned
+    """
+    if not os.path.exists(self.base_dir):
+      os.makedirs(self.base_dir)
+    self.remove_repo()
+    _, err = self._run_command(['git', 'clone', self.repo_url],
+                               self.base_dir,
+                               check_result=True)
+    if not self._is_git_repo():
+      raise RepoManagerError('%s is not a git repo' % self.repo_url)
+
+  def _run_command(self, command, location='.', check_result=False):
+    """ Runs a shell command in the specified directory location.
+
+    Args:
+      command: The command as a list to be run
+      location: The directory the command is run in
+      check_result: Should an exception be thrown on failed command
+
+    Returns:
+      The stdout of the command, the stderr of the command
+
+    Raises:
+      RepoManagerError when running a command resulted in an error
+    """
+    process = subprocess.Popen(command, stdout=subprocess.PIPE, cwd=location)
+    out, err = process.communicate()
+    if err is not None and check_result:
```

Comment:
you should also check the return code here.

Reply:
(Reopening). Can you add a check for returncode == 0 here as well, rather than just raising an exception with the return code? It's possible a process will error out without printing anything to stderr.

Also, just do `if err and check_result`. No need for the `is not None`.